### ISSUE TITLE
Potentially wrong power function gradients (opposite sign)

### ISSUE THREAD

**Issue by MtDersvan** [google/jax#6024]
Background: Jacobian and grad functions return results with the opposite sign. When checked against the Scipy grad checker the results do not satisfy near zero error requirement.

Reproducible example:

~~~python
import jax
import jax.numpy as jnp
from jax import grad, jacfwd
from jax.test_util import check_grads
from scipy.optimize import check_grad


# Scipy example:
def tfunc(x):
    return x[0]**2 - 0.5 * x[1]**3


def tgrad(x):
    return [2 * x[0], -1.5 * x[1]**2]

check_grad(tfunc, tgrad, [1.5, -1.5])
# 2.9802322387695312e-08 <- Good

check_grads(tfunc, (jnp.array([1.5, -1.5]),), order=2)
# ... <- Good

tgrad([1.5, -1.5])
# [3.0, -3.375]

grad(tfunc)([1.5, -1.5])
# [DeviceArray(3., dtype=float32), DeviceArray(-3.375, dtype=float32)]
# Nice, these results are the same, and gradients check out


# The problematic function type:
def even_power(x, alpha):
    # NOTE: really only x < 0 branch is of interest here
    return jnp.where(x < 0, jnp.power(jnp.abs(x), alpha), jnp.power(x, alpha))

def odd_power(x, alpha):
    return jnp.where(x < 0, -jnp.power(jnp.abs(x), alpha), jnp.power(x, alpha))

# Expected derivatives:
def d_even_power(x, alpha):
    return alpha*jnp.where(x < 0, jnp.power(jnp.abs(x), alpha-1.), jnp.power(x, alpha-1.))

def d_odd_power(x, alpha):
    return alpha*jnp.where(x < 0, -jnp.power(jnp.abs(x), alpha-1.), jnp.power(x, alpha-1.))


alpha = 2.0
z = jnp.array([-2.,-1.,0.,1.,2.])

d_even_power(z, alpha)
# DeviceArray([4., 2., 0., 2., 4.], dtype=float32)

d_odd_power(z, alpha)
# DeviceArray([-4., -2., 0., 2., 4.], dtype=float32)

f_even = lambda x: even_power(x, alpha)
f_odd = lambda x: odd_power(x, alpha)

[grad(f_even)(v) for v in z]
# [DeviceArray(-4., dtype=float32),
#  DeviceArray(-2., dtype=float32),
#  DeviceArray(0., dtype=float32),
#  DeviceArray(2., dtype=float32),
#  DeviceArray(4., dtype=float32)]
[grad(f_odd)(v) for v in z]
# [DeviceArray(4., dtype=float32),
#  DeviceArray(2., dtype=float32),
#  DeviceArray(0., dtype=float32),
#  DeviceArray(2., dtype=float32),
#  DeviceArray(4., dtype=float32)]

jacfwd(f_even)(z)
# DeviceArray([[-4., -0., -0., -0., -0.],
#              [-0., -2., -0., -0., -0.],
#              [ 0.,  0.,  0.,  0.,  0.],
#              [ 0.,  0.,  0.,  2.,  0.],
#              [ 0.,  0.,  0.,  0.,  4.]], dtype=float32)

jacfwd(f_odd)(z)
# DeviceArray([[4., 0., 0., 0., 0.],
#              [0., 2., 0., 0., 0.],
#              [0., 0., 0., 0., 0.],
#              [0., 0., 0., 2., 0.],
#              [0., 0., 0., 0., 4.]], dtype=float32)


# Okay, these results are an inverse to our expectations ^.
# We'd expect the odd function to have a negative mapping for negative inputs -4 and -2, and the even function - a positive one.


[check_grads(f_even, (v,), order=2) for v in z]
# [None, None, None, None, None] <- Jax thinks it's fine

[check_grad(f_even, grad(f_even), v) for v in z]
# [4.0, 2.0, 1.4901161193847656e-08, 2.0, 4.0] <- Uh-oh Scipy doesn't agree with Jax, these should be close to 0. Only v == 0 is.
~~~

If this is an expected behaviour, I'd be glad to stand corrected in my math; otherwise, would be great to see a solution.

Thanks!

**Comment by jakevdp** [comment]
In `d_even_power` and `d_odd_power`, it looks like you're not taking into account the effect of `abs` on the gradient:
```python
from jax import grad, vmap
vmap(grad(abs))(jnp.arange(-2.0, 3.0))
# DeviceArray([-1., -1.,  1.,  1.,  1.], dtype=float32)
```

**Comment by MtDersvan** [comment]
Thank you so much @jakevdp , you're absolutely right! 

Sorry, I was too busy with the actual problem I was trying to solve that I didn't notice the missing derivative. If I can get your opinion on the following issue, I would be very greatful.

The main point of abs(x) was to define f(x) on the set of negative numbers x. To do so I had 3 available options when x < 0: 1) `-x` 2) `sign(x)•x` 3) `abs(x)`. All of them slightly change the derivatives that I'd like to redefine. 
There is a percivable inconsistency with the returned results, even though these functions are supposed to turn negative x into a positive potentiation argument (if x < 0 make it -x), sometimes these functions when differentiated return all non-zero arguments into `nan`, sometimes only the negative set is `nan`, and sometimes (e.g. `jacfwd`) maps as expected. Note, `jacfwd` and `jacrev` give different results on the same inputs when using fractional powers.

~~~python
import jax.numpy as jnp
from jax import grad, jacfwd, jacrev, vmap

# Negating x < 0
def even_power_m(x, alpha):
    return jnp.where(x < 0, jnp.power(-x, alpha), jnp.power(x, alpha))

def odd_power_m(x, alpha):
    return jnp.where(x < 0, -jnp.power(-x, alpha), jnp.power(x, alpha))

# Multiplying x < 0 by its sign
def even_power_s(x, alpha):
    return jnp.where(x < 0, jnp.power(jnp.sign(x)*x, alpha), jnp.power(x, alpha))

def odd_power_s(x, alpha):
    return jnp.where(x < 0, -jnp.power(jnp.sign(x)*x, alpha), jnp.power(x, alpha))

# Using an absolute value of x
def even_power_a(x, alpha):
    return jnp.where(x < 0, jnp.power(jnp.abs(x), alpha), jnp.power(x, alpha))

def odd_power_a(x, alpha):
    return jnp.where(x < 0, -jnp.power(jnp.abs(x), alpha), jnp.power(x, alpha))

# Fixed the previously flawed differentiation
def d_even_power_a(x, alpha):
    return alpha*jnp.where(x < 0, (jnp.abs(x)/x)*jnp.power(jnp.abs(x), alpha-1.), jnp.power(x
, alpha-1.))

def d_odd_power_a(x, alpha):
    return alpha*jnp.where(x < 0, -(jnp.abs(x)/x)*jnp.power(jnp.abs(x), alpha-1.), jnp.power(
x, alpha-1.))

alpha = 1.1
z = jnp.array([-2.,-1.,0.,1.,2.])

# Define convinience wrappers
f_even_m = lambda x: even_power_m(x, alpha)
f_odd_m = lambda x: odd_power_m(x, alpha)

f_even_s = lambda x: even_power_s(x, alpha)
f_odd_s = lambda x: odd_power_s(x, alpha)

f_even_a = lambda x: even_power_a(x, alpha)
f_odd_a = lambda x: odd_power_a(x, alpha)

d_even_power_a(z, alpha), d_odd_power_a(z, alpha)
# (DeviceArray([-1.1789508, -1.1      ,  0.       ,  1.1      ,  1.1789508],            dtype=float32),
#  DeviceArray([1.1789508, 1.1      , 0.       , 1.1      , 1.1789508], dtype=float32))

vmap(grad(f_even_m))(z), vmap(grad(f_odd_m))(z)
# (DeviceArray([nan, nan,  0., nan, nan], dtype=float32),
#  DeviceArray([nan, nan,  0., nan, nan], dtype=float32))
# Why z > 0 is nan here?

 vmap(grad(f_even_s))(z), vmap(grad(f_odd_s))(z)
# (DeviceArray([      nan,       nan, 0.       , 1.1      , 1.1789509], dtype=float32),
#  DeviceArray([      nan,       nan, 0.       , 1.1      , 1.1789509], dtype=float32))

vmap(grad(f_even_a))(z), vmap(grad(f_odd_a))(z)
# (DeviceArray([      nan,       nan, 0.       , 1.1      , 1.1789509], dtype=float32),
#  DeviceArray([      nan,       nan, 0.       , 1.1      , 1.1789509], dtype=float32))

# Interestingly enough jacfwd and jacrev give different results
jacfwd(f_even_a)(z)  # Works on all functions
# DeviceArray([[-1.1789509, -0.       , -0.       , -0.       , -0.       ],
#              [-0.       , -1.1      , -0.       , -0.       , -0.       ],
#              [ 0.       ,  0.       ,  0.       ,  0.       ,  0.       ],
#              [ 0.       ,  0.       ,  0.       ,  1.1      ,  0.       ],
#              [ 0.       ,  0.       ,  0.       ,  0.       ,  1.1789509]],            dtype=float32)

jacrev(f_even_a)(z)
# DeviceArray([[      nan,       nan, 0.       , 0.       , 0.       ],
#              [      nan,       nan, 0.       , 0.       , 0.       ],
#              [      nan,       nan, 0.       , 0.       , 0.       ],
#              [      nan,       nan, 0.       , 1.1      , 0.       ],
#              [      nan,       nan, 0.       , 0.       , 1.1789509]],            dtype=float32)
~~~

The questions are:
- Why the differentiation returns inconsistent results? How to make them consistent?
- Is it possible to override the way gradients for these function are calculated?

Best Regards

**Comment by jakevdp** [comment]
It sounds like you might be hitting some version of this issue: https://jax.readthedocs.io/en/latest/faq.html#gradients-contain-nan-where-using-where

**Comment by MtDersvan** [comment]
This is perfect, thanks. Again, sorry for bothering.
